### ISSUE TITLE
exec: "auplink": executable file not found in $PATH

### ISSUE THREAD

**Issue by tianon** [docker-library/docker#25]
Currently (at least, as of 2016-09-23), `auplink` isn't available in Alpine (https://pkgs.alpinelinux.org/contents?file=auplink), so for us to include it, we'd have to compile it from source (https://sourceforge.net/p/aufs/aufs-util/).

As I understand it, its use within Docker is primarily for detecting and properly handling hard links between layers, and beyond that particular use case not having it available should be mostly harmless.  I'm opening this issue (in spite of [already having a `TODO` in the `Dockerfile` itself](https://github.com/docker-library/docker/blob/64907cf604ea5dad5eb22f9b3f65fab6f97f05fb/1.12/dind/Dockerfile#L12)) for tracking/conversation purposes, and to help interested/concerned travellers discover more information about why they're seeing this error in their Docker-in-Docker containers.

So, TL;DR, we don't have `auplink` because there isn't an Alpine package for it (yet?), but for most users that's harmless anyhow. :+1:

**Comment by floatdrop** [comment]
We bumped on this issue with docker-dind version 1.11.2:

```level=error msg="Couldn't run auplink before unmount /var/lib/docker/tmp/docker-aufs-union030413474: exec: \"auplink\": executable file not found in $PATH"```

Is there any way to workaround it?

**Comment by tianon** [comment]
@floatdrop are you actually seeing any issues related to the error?

**Comment by floatdrop** [comment]
@tianon it is hard to say, after some investigation I'm think this error was always there (but hidden in disabled debug output). We are using dind from [drone docker](https://github.com/drone-plugins/drone-docker) plugin and it does not provide much useful output, when docker failing to build and publish image. All that I'm seeing is exit code `1`. We will dig further.

I'm right, that `aufs-tools` was excluded long time ago?

**Comment by tianon** [comment]
It's actually _never_ been included in the `docker:dind` image (so it's unlikely to be the cause of your issues). :confused:

**Comment by tianon** [comment]
:open_mouth: Alpine Edge appears to have `aufs-util`! (https://pkgs.alpinelinux.org/contents?file=auplink)

So, we _might_ be able to fix this once Alpine 3.7 is released. :smile:

**Comment by phedoreanu** [comment]
Until 3.7 is released you can install `aufs-util` running:

`apk add -U --repository http://dl-cdn.alpinelinux.org/alpine/edge/testing aufs-util`.

**Comment by tianon** [comment]
Aw, `aufs-util` is still in Edge (it wasn't included in either 3.7 _or_ 3.8).

On the flip side, both [Docker for Windows][1] and [Docker for Mac][2] have deprecated AUFS, so perhaps we can give up on AUFS and stop worrying about it too? :smile:

[1]: https://docs.docker.com/docker-for-windows/release-notes/#docker-community-edition-18060-ce-win70-2018-07-25
[2]: https://docs.docker.com/docker-for-mac/release-notes/#docker-community-edition-18060-ce-mac70-2018-07-25

**Comment by tao12345666333** [comment]
@tianon  Maybe we can. We have more modern chosen `overlay` or `overlay2`.

In the [docker document](https://docs.docker.com/storage/storagedriver/aufs-driver/), also has suggestions.

>  If your Linux kernel is version 4.0 or higher, and you use Docker CE, consider using the newer overlay2, which has potential performance advantages over the aufs storage driver.

**Comment by tianon** [comment]
Good point, I'd forgotten that Docker's preference order changed such that `overlay2` is higher now (and has been for quite a while). :+1:

So, given the following, I'm going to close this now: :taco:

- https://docs.docker.com/docker-for-windows/release-notes/#docker-community-edition-18060-ce-win70-2018-07-25
- https://docs.docker.com/docker-for-mac/release-notes/#docker-community-edition-18060-ce-mac70-2018-07-25
- https://docs.docker.com/storage/storagedriver/aufs-driver/
- https://github.com/docker/docker/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/daemon/graphdriver/driver_linux.go#L50-L51

**Comment by tianon** [comment]
https://github.com/moby/moby/pull/38090 is a new PR that's relevant here -- officially deprecating AUFS in 19.03+ :metal:
